gnu/JackAtomic_os.h

```diff
@@ -0,0 +1,31 @@
+/*
+Copyright (C) 2004-2008 Grame
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+
+*/
+
+#ifndef __JackAtomic_gnu__
+#define __JackAtomic_gnu__
+
+#include "JackTypes.h"
+
+static inline char CAS(volatile UInt32 value, UInt32 newvalue, volatile void* addr)
+{
+    return __sync_bool_compare_and_swap ((UInt32*)addr, value, newvalue);
```

Comment:
I don't think these ancient GCC `__sync_*`  builtins should be used in new code at all. In general jack2's atomic code is pretty bad by modern standards; it's full of sus `volatile` sprinkled everywhere and relies on inline asm, making it janky and most likely slower than it should (thanks to the cache flushing caused by `volatile`).

Both C and C++ have really good standard atomics support that would serve the purpose much better, but that would be out of scope for this PR. IMHO at the very least this should use the slightly less outdated `__atomic_*` GCC builtins, assuming Clang also supports those. Can't remember. It wouldn't be as nice as `std::atomic` but a step in the right direction.

Reply:
`volatile` by itself doesn't imply cache flushing. But synchronization memory coherency properties do indeed. 

What is CAS() supposed to provide as property? `__atomic_*` indeed allows to specify something more precise than a full barrier, but without knowing what it should be (and from a quick look to the existing code it seems it's a full barrier that is needed), using `__atomic_*` won't be any better than the linux port's `__sync_*` that I replicated for coherency. If I start using `__atomic_*` in the gnu port, somebody who will be fixing all these os-specific pieces might start wondering why the gnu port is different from the linux port. Is that really what we want?

Reply:
> using __atomic_* won't be any better than the linux port's __sync_* that I replicated for coherency

(Apparently  `__atomic_*` would allow to drop the memory barrier on comparison failure. In the cases where `CAS` is used (while loops until success) that won't change anything).

> that I replicated for coherency

What I also fear is that if I make the gnu port depart from the strong guarantee of the linux version, people might start producing code that actually require the strong guarantee and work on linux but be bogus on the gnu port.

Reply:
To make progress at all, I have updated to use the `__atomic_*` equivalent of `__sync_*`